unicodetools/src/main/java/org/unicode/xml/UcdXML.java

```diff
@@ -0,0 +1,825 @@
+                qName = qName.substring(2);
+            }
+            String orgCharAttributesValue = orgRangeAttributes.getValue(qName);
+            String groupAttributeValue = groupAttrs.getValue(qName);
+            if (!Objects.equals(orgCharAttributesValue, groupAttributeValue)) {
+                rangeAttributes.addAttribute(
+                        NAMESPACE,
+                        qName,
+                        qName,
+                        "CDATA",
+                        Objects.requireNonNullElse(orgCharAttributesValue, ""));
+            }
+        }
+        writer.startElement(rangeType.tag, rangeAttributes);
+        {
+            writer.endElement(rangeType.tag);
+        }
+    }
+
+    private static void buildUngroupedRange(
+            UCDXMLWriter writer,
+            AttributeResolver attributeResolver,
+            VersionInfo ucdVersion,
+            ArrayList<Integer> range,
+            Range rangeType)
+            throws SAXException {
+        AttributesImpl rangeAttributes =
+                getReservedAttributes(ucdVersion, attributeResolver, range);
+        writer.startElement(rangeType.tag, rangeAttributes);
+        {
+            writer.endElement(rangeType.tag);
+        }
+    }
+
+    private static boolean isWritableCodepoint(
+            int codepoint, UCDXMLOUTPUTRANGE outputRange, AttributeResolver attributeResolver) {
+        return outputRange == UCDXMLOUTPUTRANGE.ALL
+                || (outputRange == UCDXMLOUTPUTRANGE.UNIHAN
+                        && attributeResolver.isUnihanAttributeRange(codepoint))
+                || (outputRange == UCDXMLOUTPUTRANGE.NOUNIHAN
+                        && !attributeResolver.isUnifiedIdeograph(codepoint));
+    }
+
+    private static Range getRangeType(AttributeResolver attributeResolver, int codepoint) {
+        String NChar = attributeResolver.getNChar(codepoint);
+        UcdPropertyValues.General_Category_Values gc = attributeResolver.getgc(codepoint);
+
+        if (attributeResolver.isUnihanAttributeRange(codepoint)) {
+            return Range.CJKUNIFIEDIDEOGRAPH;
+        }
+        if (gc.equals(UcdPropertyValues.General_Category_Values.Surrogate)) {
+            return Range.SURROGATE;
+        }
+        if (gc.equals(UcdPropertyValues.General_Category_Values.Private_Use)) {
+            return Range.CHARACTER;
+        }
+        if (NChar.equals(UcdPropertyValues.Binary.Yes.getShortName())) {
+            return Range.NONCHARACTER;
+        }
+        return Range.RESERVED;
+    }
+
+    private static int getLastCodepointInGroup(
+            AttributeResolver attributeResolver, int lowCodepoint, int highCodepoint) {
+        String blk = attributeResolver.getAttributeValue(UcdProperty.Block, lowCodepoint);
+        for (int codepoint = lowCodepoint; codepoint <= highCodepoint; codepoint++) {
+            if (!blk.equals(attributeResolver.getAttributeValue(UcdProperty.Block, codepoint))) {
```

Comment:
nit: it should be easy to get the last code point of the blk from the IUP, and do that once outside the loop